### DISCUSSION TITLE
Mutations and Returning Data Types

### DISCUSSION THREAD

**Discussion by Vyse777** [Netflix/dgs-framework#1141] [Q&A]
Hi team,
I am having trouble understanding how mutations operate within the DGS framework.

When a client wishes to initiate a mutation, how will/does DGS operate when the return type is a large entity that might be comprised of multiple DGS data fetchers to get all of its properties?
Does it run a query for the entity after the mutation? If so how does it know?
Or, is it the duty of the mutation method to return the entire entity? If so how the heck would you do this within the framework?

Example for discussion:
Schema:
```
type Query {
   largeThing(Id: String): LargeThing
}

type Mutation {
   updateThingValueC(thingId: String, cValue: String): LargeThing # Note that this is the entire type as the response. Not just the single thing (String) this mutation is allowing to be updated
}

type LargeThing {
   Id: String
   a: String # Say for argument this field comes from a more complex query
   b: String # Say for argument this field comes from an entirely different system and so has it's own data loader for queries
   c: String # This is the field we wish to mutate
.....
}
```

In short - I want to make a mutation that only updates the state of a single field on this `LargeThing`, but want to expose the opportunity to the client to get back the entire entity as the "response" to this mutation. 
How would this work? 
Is it possible within the capabilities of the framework? 
Or, am I even completely off base on this being a 'standard' GraphQL way of doing something like this (which is totally possible being fairly new to GraphQL still)?

I am really enjoying using this framework so far (in Kotlin) and would love to know how best to do this. Please let me know! And if you have any follow up questions or need elaboration, I will do my best to help.

Thanks!

**Comment by srinivasankavitha** [comment]
Your data fetcher for the mutation will handle the logic for updating the single field. Since you will need access to `LargeThing` instance to make that update, I assume you will be making a call to fetch the required object anyway. As long as you specify the return type as the `LargeThing`, you can simply return that in your data fetcher for `updateThinkValueC`.

Hope that answers your question.

**Reply by Vyse777** [reply]
Mmm not quite.

When a client says they want `LargeThing` and all if its fields from the response of the mutation - these fields might be spread across multiple microservices as a part of a "federated type", right?

If that is the case, it's not unfeasible to assume that a mutation that returns the entire entity AND it's federated sub-models, would require orchestration across multiple of the microservices that comprise a particular entity, in this case `LargeThing`

So how does someone, from the context of a single microservice, fetch the entire entity utilizing the framework? Or even become aware of if the client is even asking for specific fields???
Is that left up to the developer (me) or is there a mechanism provided by this framework that enables this?

For a more familiar example lets use the Shows and Reviews schema:
For this discussion, lets say that `Shows` are owned my the Shows-Microservice, and `Reviews` are owned by the Reviews-Microservice. Two independent systems deployed in the ether.

The Shows microservice schema might look like this:
```
type Query {
  shows(titleFilter: String): [Show]
  show(Id: String): Show
}

# NOTE: A federated type
type Show @key(fields: "id") { 
  id: ID
  title: String
  releaseYear: Int
}
```

And the Reviews Microservice schema might look like this:
```
# NOTE: Extending the federated type
type Show @key(fields: "id") @extends {
  id: ID @external
  reviews: [Review]
}

type Review {
  starRating: Int
}
```

Comprising a merged schema in Apollo like so:
```
type Query {
  shows(titleFilter: String): [Show]
  show(Id: String): Show
}

type Show @key(fields: "id") {
  id: ID
  title: String
  releaseYear: Int
  reviews: [Review] # From the Reviews microservice
}

type Review {
  starRating: Int
}
```

Now lets throw a mutation in there that does something like this, NOTE: this would be defined in the Shows Microservice and NOT the Reviews Microservice:
```
type Mutation {
  updateShowReleaseYear(showId: String, newReleaseYear: Int): Show # NOTE: It allows the client to fetch the whole Show model, in this case including the Reviews!
}
```

With this, as it is transparent to the client which microservice does what (i.e. the client doesn't know or care), if the client says "update a shows release year with id 3 to 2022" and it requests the Show WITH REVIEWS from the response of the mutation. How would you do that with the tools in this framework???

From within the context of the Shows Microservice, that is implementing the mutation for "Update this Show", it has no knowledge of the "Reviews" the mutation says it can return, nor would I expect it to! Does the Shows Microservice need to communicate with the Reviews Microservice in order to solve this problem? If this is the case, federated types are severely limited... In essence this would eliminate the decoupling federation provides.

Does this clear up my question any better? Please let me know if you need me to clarify further. I appreciate you taking the time to review my question.

**Reply by srinivasankavitha** [reply]
Ok, I understand what you are asking. This is where you take advantage of federation. With federation, you are required to have both the Shows and Reviews service behind a federated gateway that is aware of the individual DGS schemas. It is the responsibility of this gateway to  contact the Shows service for Show fields and Reviews service to fetch the reviews by passing along the showId. The reviews service does not need to populate any other fields in the `Show` object it is returning, just the `reviews`. The gateway will merge the responses from Show service and Review Service and send it back to the client.

At Netflix, we have our own internal implementation of the federated gateway, but the DGS framework works just as well with Apollo's gateway offering.

I hope that answers your question. I'm also linking a few resources that go into the concepts in more detail:
https://www.apollographql.com/docs/federation/federation-spec/

**Reply by Vyse777** [reply]
Still not quite. But at least we are on the same page with federation!

With this example (or at least what initiated this original question) I am using Apollo for federation. So that is good.

How does federation tie into the mutation example discussed above now?

When a client initiates a mutation on `Show` (ex from above: `updateShowReleaseYear`) , but asks for `ShowTitle`, and `Review` data in response, how does federation operate to fetch the Review data? Again, from the context of the mutation being handled by the Shows Microservice it has no knowledge of that data.

Is Apollo smart enough to know it has to initiate a Query to the Reviews Microservice even though the initial request was a Mutation?
From my experiments, this does not seem to be the case.

**Reply by srinivasankavitha** [reply]
This should be handled by the gateway for the scenario you describe as well. I just tried it with the Apollo gateway and it works. Be sure to define a `@DgsEntityFetcher` for Show on both the Shows service and Reviews service. That is how the gateway will contact each service for fetching data corresponding to the query and the DGS schema. Here is some basic code wiring up the entity fetcher for the Shows service:
```
    @DgsMutation
    fun updateShow(@InputArgument showId: String?) : Show {
       // TODO: add custom logic to mutate
        return showsService.shows().filter { it.id == showId }[0]
    }

    @DgsEntityFetcher(name = "Show")
    fun show(entityKeys: Map<String, Any>) : Show {
        val showId = entityKeys["id"]
        return showsService.shows().filter { it.id == showId }[0]
    }
```

Hope that clarifies things better.

**Reply by Vyse777** [reply]
To be clear this is returning the show entity on the mutation, and then you are saying that Apollo is calling the Reviews Microservice to fetch `Reviews` even though it's part of a mutation response and not a query? I have no luck getting this to work!..

**Reply by srinivasankavitha** [reply]
Yes that is correct, and the basis of how federation should work.  It is important that the Reviews service has the @DgsEntityFetcher implemented and returns a proper `Show` object with the reviews field populated.

**Reply by Vyse777** [reply]
Just a followup - I am still planning on looking into this. I haven't had the time to experiment with it.

Again thanks for your time!
